animation-proposal.md

```diff
@@ -0,0 +1,338 @@
+- Unity's animation system cannot operate in app-relative coordinates.
+- Our API for programmatic generation of keyframed animations is both unwieldly and of limited use.
+- Our existing `animateTo` implementation leverages the current animation system, which is bad.
+
+
+
+
+## Proposal
+
+It isn't feasable to implement our own animation system, but I also believe we don't need to. I instead recommend we support playing native animations, and we implement a simple but powerful interpolation API that supports interpolating any mutable property on `Actor` or `Asset`.
+
+In short:
+- Add support for playing animations defined on assets (loaded from glTF files, Unity asset bundles, etc.).
+- Add generalized support for interpolating any mutable property of an `Actor` or `Asset`.
+- Remove support for programmatic animaton generation.
+- Remove the existing `animateTo` call in favor of the new interpolation solution.
+
+
+
+## API
+
+### Animations
+
+#### Move animation API to new class called `Animation`
+
+When a model is loaded from a resource, its animations will be included in the `object-spawned` reply, and be available on `Actor` as an array of `AnimationLike` types. `AnimationLike` will define the API for starting, stopping, seeking, etc. animations.
+
+```ts
+export interface AnimationLike {
+	/**
+	 * getter that returns the name of the animation.
+	 */
+	public name: string;
+
+	/**
+	 * Starts playing the animation from the beginning.
+	 * @returns A Promise that resolves when the animation completes or is stopped.
+	 */
+	public start(): Promise<void>;
+
+	/**
+	 * Stops the animation on the actor and resolves promises returned from `completed()`.
+	 */
+	public stop();
+
+	/**
+	 * Pauses the animation (sets animation speed to zero).
+	 */
+	public pause();
+
+	/**
+	 * Resumes the animation (sets animation speed to 1).
+	 */
+	public resume();
+
+	/**
+	 * Sets the animation time (units are in seconds).
+	 */
+	public setTime(time: number);
+
+	/**
+	 * Sets the playback speed.
+	 */
+	public setSpeed(speed: number);
+
+	/**
+	 * (Advanced) Sets the time, speed, and enabled state of an animation.
+	 */
+	public setState(state: SetAnimationStateOptions);
+
+	/**
+	 * @returns A Promise that resolves when the animation completes or is stopped.
+	 */
+	public completed(): Promise<void>;
+}
+```
+
+Then, on `Actor`:
+
+```ts
+/**
+ * getter that returns the array of animations defined on this actor.
+ */
+public get animations(): AnimationLike[];
+
+/**
+ * Get an animation by name.
+ */
+public animation(animationName: string): AnimationLike;
+```
+
+#### Examples
+
+```ts
+// Play the 'jump' animation, then do something when it's done.
+actor.animation("jump").start().then(() => console.log("done jumping"));
+
+// Play the 'victory' animation, and wait for it to complete.
+await actor.animation("victory").start();
+```
+
+### Interpolations
+
+#### Add `Context.interpolate` method
```

Comment:
Context seems like an awkward place to put this. Util method? Or maybe a static method on the Animation class?

Reply:
Agreed. I mention below the idea of adding it to both Actor and Asset, to make it more discoverable. Note that interpolations are no longer related to animations in this new system, so it wouldn't make sense to add it to Animation.

Reply:
Added to Actor and Asset, and removed from Context.

Reply:
Actually, there are some issues with moving this method off Context. Reverting that design change for the moment.